### ISSUE TITLE
Rendering problem with Mapbox custom layer

### ISSUE THREAD

**Issue by ThisIsBarney** [uber/react-map-gl#735]
**Version**: 4.0.11
**Problem**: When using custom layer in Mapbox, after I dragged the map, the render method (shown below) will **no longer be triggered** until I start to drag the map again.
```
class CustomLayer {
  constructor() {
    this.id = 'custom_layer';
    this.type = 'custom';
    this.renderingMode = '3d';

    // xxxxxxx
  }

  onAdd = (map: mapboxgl.Map, gl: WebGLRenderingContext) => {
    // some initialization
  }

  render = (gl: WebGLRenderingContext, matrix: number[]) => {
    // nothing will appear here after drag the map
  }
}

onViewportChange = (viewport) => {
    this.setState({
      viewport: {
        ...this.state.viewport, ...viewport
      },
    })
  }

render() {
    const {
      viewport,
    } = this.state;
    return (
      <div>
        <MapGL
          {...viewport}
          mapStyle={MAP_STYLE}
          onViewportChange={this.onViewportChange}
          ref={ref => { this.mapbox = ref; }}
        >
        </MapGL>
      </div>
    );
  }
```

BTW, there is no problem with Version 4.0.10.

**Comment by Pessimistress** [comment]
@ThisIsBarney I am unable to reproduce this behavior using your code sample. `render` is still triggered. Can you provide more context of your use case? Do you have animations in the custom layer?

Between 4.0.10 and 4.0.11 we introduced a change where we redraw the map synchronously inside the React render cycle. More context [here](https://github.com/mapbox/mapbox-gl-js/issues/7893).

**Comment by ThisIsBarney** [comment]
@Pessimistress Yes, I have animations in the custom layer to update 3d Objects. I use three.js in custom layer. The code of animation is like:
```
animate = () => {
  // do some transformations of the three.js objects
  
  window.requestAnimationFrame(this.animate);
}
```
`render` behaves correct if I don't drag the map, but after I dragged the map and stopped dragging, the `render` stops. The `animate` function still get called every frame when `render` stops.

**Comment by Pessimistress** [comment]
Are you using `triggerRepaint` to update the frame?

**Comment by ThisIsBarney** [comment]
@Pessimistress YES, I use `triggerRepaint` in render function of the custom layer class

**Comment by Pessimistress** [comment]
OK that makes sense now. I can fix this.
Though I still think using `triggerRepaint` inside a render function is weird. Where is the `animate` function living then?

**Comment by ThisIsBarney** [comment]
`animate` lives in `App` scope (not in the custom layer class) and do some schedule tasks (include transfrom 3d objects) while `triggerRepaint` lives in the `render` function of the custom layer class. 

Any suggestions would be greatly appreciated lol.

**Comment by Pessimistress** [comment]
Well on a high level, whoever updates the animation should trigger the repaint. Imagine your animation does not update every frame (e.g. a transition that ends after 2 seconds), your approach will never stop rerendering, and would be very expensive from the performance perspective.

Assuming `App` renders `ReactMap`:

```js
animate = () => {
  // update animation

  this._reactMapRef.getMap().triggerRepaint();
  
  window.requestAnimationFrame(this.animate);
}
```

**Comment by ThisIsBarney** [comment]
thanks, trigger in animate solves the problem!

**Comment by Pessimistress** [comment]
Render bug fixed in 4.0.12.
